dataRetriever/resolvers/equivalentProofsResolver.go

```diff
@@ -0,0 +1,293 @@
+	}
+	if check.IfNil(args.EquivalentProofsPool) {
+		return dataRetriever.ErrNilProofsPool
+	}
+
+	return nil
+}
+
+// ProcessReceivedMessage represents the callback func from the p2p.Messenger that is called each time a new message is received
+// (for the topic this validator was registered to, usually a request topic)
+func (res *equivalentProofsResolver) ProcessReceivedMessage(message p2p.MessageP2P, fromConnectedPeer core.PeerID, source p2p.MessageHandler) ([]byte, error) {
+	err := res.canProcessMessage(message, fromConnectedPeer)
+	if err != nil {
+		return nil, err
+	}
+
+	res.throttler.StartProcessing()
+	defer res.throttler.EndProcessing()
+
+	rd, err := res.parseReceivedMessage(message, fromConnectedPeer)
+	if err != nil {
+		return nil, err
+	}
+
+	switch rd.Type {
+	case dataRetriever.HashType:
+		return nil, res.resolveHashRequest(rd.Value, rd.Epoch, message.Peer(), source)
+	case dataRetriever.HashArrayType:
+		return nil, res.resolveMultipleHashesRequest(rd.Value, rd.Epoch, message.Peer(), source)
+	case dataRetriever.NonceType:
+		return nil, res.resolveNonceRequest(rd.Value, rd.Epoch, message.Peer(), source)
+	default:
+		err = dataRetriever.ErrRequestTypeNotImplemented
+	}
+	if err != nil {
+		return nil, fmt.Errorf("%w for value %s", err, logger.DisplayByteSlice(rd.Value))
+	}
+
+	return []byte{}, nil
+}
+
+// resolveHashRequest sends the response for a hash request
+func (res *equivalentProofsResolver) resolveHashRequest(hashShardKey []byte, epoch uint32, pid core.PeerID, source p2p.MessageHandler) error {
+	headerHash, shardID, err := getHashAndShard(hashShardKey)
+	if err != nil {
+		return fmt.Errorf("resolveHashRequest.getHashAndShard error %w", err)
+	}
+
+	data, err := res.fetchEquivalentProofAsByteSlice(headerHash, shardID, epoch)
+	if err != nil {
+		return fmt.Errorf("resolveHashRequest.fetchEquivalentProofAsByteSlice error %w", err)
+	}
+
+	return res.marshalAndSend(data, pid, source)
+}
+
+// resolveMultipleHashesRequest sends the response for multiple hashes request
+func (res *equivalentProofsResolver) resolveMultipleHashesRequest(hashShardKeysBuff []byte, epoch uint32, pid core.PeerID, source p2p.MessageHandler) error {
+	b := batch.Batch{}
+	err := res.marshalizer.Unmarshal(&b, hashShardKeysBuff)
+	if err != nil {
+		return err
+	}
+	hashShardKeys := b.Data
+
+	equivalentProofsForHashes, err := res.fetchEquivalentProofsSlicesForHeaders(hashShardKeys, epoch)
+	if err != nil {
+		return fmt.Errorf("resolveMultipleHashesRequest.fetchEquivalentProofsSlicesForHeaders error %w", err)
+	}
+
+	return res.sendEquivalentProofsForHashes(equivalentProofsForHashes, pid, source)
+}
+
+// resolveNonceRequest sends the response for a nonce request
+func (res *equivalentProofsResolver) resolveNonceRequest(nonceShardKey []byte, epoch uint32, pid core.PeerID, source p2p.MessageHandler) error {
+	data, err := res.fetchEquivalentProofFromNonceAsByteSlice(nonceShardKey, epoch)
```

Comment:
i think we can keep all proofs into one map for all shards, so that we don't have to consider shardId here, it's not really needed (by shard) in proofs pool (we should have some benchmarks for this)